geo/geopoint.go

```diff
@@ -57,25 +75,43 @@ func NewGeoPoint(lat, lon float64) (*Coordinate, error) {
 }
 
 // CircleToPolygon generates a bounding box (circle) from a single geopoint using radius (in metres) and the number of segments
-func (geoPoint *Coordinate) CircleToPolygon(radius float64, segments int) (*GeoStructure, error) {
+func (geo *Config) CircleToPolygon(geoPoint Coordinate, radius float64, segments int) (*GeoStructure, error) {
 
 	// validate input
-	if err := validateInput(geoPoint, radius, segments); err != nil {
+	if err := geo.validateInput(&geoPoint, radius, segments); err != nil {
 		return nil, err
 	}
 
 	shape := &GeoStructure{
 		Type: polygon,
 	}
 
-	var coordinates [][]float64
+	coordinates := make([][]float64, segments)
+
+	var semaphoreChan = make(chan struct{}, geo.defaultConcurrencyLimit)
+
+	var wg sync.WaitGroup
 
 	for i := 0; i < segments; i++ {
-		segment := (twoPi * float64(-i)) / float64(segments)
-		coordinate := generateCoordinate(*geoPoint, radius, segment)
-		coordinates = append(coordinates, coordinate)
+
+		semaphoreChan <- struct{}{}
```

Comment:
(question): Not sure if I get the logic right here, you have a semaphoreChan buffered channel with defaultConcurrencyLimit. But then you are already spinning go routines which equals the number of segments. Now are you implying that even though you have started n number of goroutines, the number of goroutines which will be  working in parallel to generate coordinate will be equal to the default concurrency limit ? if so we could start only that many number of goroutines in the first place right, why do we have more goroutines started ?

Reply:
Yes you are right, I remember thinking this too, but never refactored it. I will use the lowest value between the `geo.defaultConcurrencyLimit` and the number of `segments` passed into method when buffering the number of go routines